src/Elastic.Apm/Metrics/Windows/GlobalMemoryStatusEx.cs

```diff
@@ -0,0 +1,35 @@
+﻿using System.Runtime.InteropServices;
+
+namespace Elastic.Apm.Metrics.Windows
+{
+	[StructLayout(LayoutKind.Sequential)]
+	internal struct MemoryStatusEx
+	{
+		internal uint dwLength;
+		private readonly uint dwMemoryLoad;
+		internal readonly ulong ullTotalPhys;
+		internal readonly ulong ullAvailPhys;
+		private readonly ulong ullTotalPageFile;
+		private readonly ulong ullAvailPageFile;
+		private readonly ulong ullTotalVirtual;
+		private readonly ulong ullAvailVirtual;
+		private readonly ulong ullAvailExtendedVirtual;
+	}
+
+	internal static class GlobalMemoryStatus
+	{
+		[return: MarshalAs(UnmanagedType.Bool)]
+		[DllImport("Kernel32.dll", CharSet = CharSet.Auto, SetLastError = true)]
+		private static extern bool GlobalMemoryStatusEx(ref MemoryStatusEx lpBuffer
+		); //ref. vs. [In, Out], see: https://www.pinvoke.net/default.aspx/kernel32.globalmemorystatusex
+
+		internal static (bool success, ulong total, ulong avail) GetTotalPhysAndAvailPhys()
+		{
+			if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows)) return (false, 0, 0);
```

Comment:
I think it's over-defensive programming - if this method is called when running under OS that is not Windows then the calling code should be fixed. IMHO defensive programming is there to establish safe perimeter against uncertain run-time conditions but it should not swallow bugs.

Reply:
I think you can never be over-defensive when a bug in our code can crash the production system of a user. 

If this would be in another context, with not having the requirement of "never ever ever crash the user's application", I'd let it throw an exception. But not with this requirement.

Regarding "swallowing bug": if this is called on a non windows OS then we don't return valid data, and the effect is that we don't send memory usage info - that won't hurt any application. If we take the other approach and let this throw then the consequence is that we let a user application crash. I'd much rather "swallow this bug" than let it crash the app.

Reply:
I agree agent's should never disrupt application and defensive programming should be used but it should be systematic and limited to the boundary between the agent and the application instead of randomly sprinkled over agent's internal utils code. 

The concern of agent disrupting application should be solved on much higher level than agent's internal utils code. As seemingly unrelated example - JDK 1.0 had all collection classes thread safe until it became clear that it doesn't contribute much to application's thread safety and just causes performance overhead - thread safety should be addressed on much higher level than collection's library and so JDK's new collections library had thread unsafe collection as well. The same applies here - defensive programming should be applied on the boundary. Unfortunately we didn't do it yet in a systematic fashion but in this case you did do it - `MetricsCollector` already captures any exception escaping from a metrics provider. I think it's also worth wrapping the whole `MetricsCollector.CollectAllMetrics` to prevent exception escaping from timer callback. I'll add a separate comment for that. 

If we take your approach wouldn't we need to check every argument for null before accessing it in every method? I think input validation should be done on boundary where unverified input gets into the component but repeating it in internal code is just a sign that we don't trust systematic handling at the boundary.

Reply:
I just remove that line, we spend too much time on this.